streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamUtil.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+
+/**
+ * Shared functions to handle verifications of a valid {@link org.apache.kafka.streams.kstream.KStream}.
+ */
+final class KStreamUtil {
+
+    private KStreamUtil() {}
+
+    /**
+     * @throws IllegalArgumentException if the same transformer instance is obtained each time
+     */
+    static void checkTransformerSupplier(final TransformerSupplier<?, ?, ?> supplier) {
+        if (supplier.get() == supplier.get()) {
+            throw new IllegalArgumentException("TransformerSupplier generates single transformer reference. Supplier " +
+                    "pattern violated.");
```

Comment:
I think we can be more elaborate and add something like `TransformerSupplier#get() must return a new Transformer object each time it is called.`